sparrow_datums/keypoints/keypoints_test.py

```diff
@@ -0,0 +1,165 @@
+        PType.absolute_xy,
+        image_width=image_size,
+        image_height=image_size,
+    )
+    keypoints_b = keypoints_a.to_relative()
+    assert keypoints_b.is_relative
+    np.testing.assert_array_less(keypoints_b.array, 1.0)
+    keypoints_c = keypoints_b.to_relative()
+    assert keypoints_c.is_relative
+
+
+def test_to_absolute_moves_keypoints_to_0_image_size():
+    image_size = 512
+    keypoints_a = Keypoints(
+        np.random.uniform(size=(10, 2)),
+        PType.relative_xy,
+        image_width=image_size,
+        image_height=image_size,
+    )
+    keypoints_b = keypoints_a.to_absolute()
+    assert keypoints_b.is_absolute
+    assert (keypoints_b.array < 1).mean() < 0.1
+    keypoints_c = keypoints_b.to_absolute()
+    assert keypoints_c.is_absolute
+
+
+def test_keypoint_slicing_cols_throws_value_error():
+    keypoints = Keypoints(np.random.uniform(size=(10, 2)))
+    with pytest.raises(ValidationError):
+        keypoints[:, :1]
+    top_left = keypoints.array[:, :1]
+    assert isinstance(top_left, np.ndarray)
+
+
+def test_keypoints_deserializes_type():
+    keypoints = Keypoints.from_dict(
+        {
+            "data": [0.2, 0.3],
+            "ptype": "relative_xy",
+            "image_width": None,
+            "image_height": None,
+            "fps": None,
+            "object_ids": None,
+        }
+    )
+    assert keypoints.ptype == PType.relative_xy
+
+
+def test_keypoints_saves_classname():
+    data = Keypoints(np.ones(2)).to_dict()
+    assert data["classname"] == "Keypoints"
+
+
+def test_keypoint_attributes_require_known_keypoint_parameterization():
+    with pytest.raises(ValidationError):
+        Keypoints(np.ones(2)).x
+
+
+# def test_creating_heatmap():
+#     image_size = 512
+#     keypoints = Keypoints(
+#         np.random.uniform(size=(10, 2)),
+#         PType.relative_xy,
+#         image_width=image_size,
+#         image_height=image_size,
+#     )
+#     heatmaps_array = keypoints.to_heatmap_array()
+#     assert heatmaps_array.shape == (10, image_size, image_size)  # check the shape
+#     assert type(heatmaps_array) is np.ndarray
+#     assert np.min(heatmaps_array) == 0
+#     assert np.max(heatmaps_array) == 1
+#     heatmaps = Heatmaps(data=heatmaps_array, ptype=PType.heatmap)
+#     assert type(heatmaps) is Heatmaps
+#     # change the co-variance
+#     heatmaps_array = keypoints.to_heatmap_array(covariance=10)
+#     np.testing.assert_array_less(heatmaps_array, 1.0000001)
+
+
+def test_back_and_forth_between_keypoints_and_heatmaps_for_1D_arrays():
+    keypoint = Keypoints(
+        np.array([4, 8]),
+        PType.absolute_xy,
+        image_height=21,
+        image_width=28,
+    )
+    heatmap_array = keypoint.to_heatmap_array()
+    assert heatmap_array.shape == (keypoint.image_height, keypoint.image_width)
+    assert np.max(heatmap_array) == 1
+    assert np.min(heatmap_array) == 0
+    back_to_keypoint = Keypoints.from_heatmap_array(heatmaps=heatmap_array)
```

Comment:
@jbencook, A quick goofy question if you could answer. Line 118 throws me the following error. 
`FAILED sparrow_datums/keypoints/keypoints_test.py::test_back_and_forth_between_keypoints_and_heatmaps_for_1D_arrays - TypeError: from_heatmap_array() missing 1 required positional argument: 'self'`
The reason is that it should come from an instantiated object such as the `keypoint` at line 108. My question is that if we needed `keypoint` at 108 for this method to work; why don't we just do `keypoint.array` to get the same result without the hassle of having a dedicated method for it?

Reply:
Yeah, the method should be a `classmethod`: https://www.tutorialsteacher.com/python/classmethod-decorator

That will solve the "self" exception you're getting.